services/tns_submission_queue/tns_submission_queue.py

```diff
@@ -35,425 +44,613 @@
 report_url = urllib.parse.urljoin(TNS_URL, 'api/bulk-report')
 search_frontend_url = urllib.parse.urljoin(TNS_URL, 'search')
 
-queue = []
-
-
-def tns_submission(
-    obj_ids,
-    tnsrobot_id,
-    user_id,
-    reporters="",
-    archival=False,
-    archival_comment="",
-    instrument_ids=[],
-    stream_ids=[],
-    parent_session=None,
-):
-    """Submit objects to TNS.
-    obj_ids : List[str]
-        Object IDs
-    tnsrobot_id : int
-        TNSRobot ID
-    user_id : int
-        SkyPortal ID of User posting to TNS
-    reporters : str
-        Reporters to appear on TNS submission.
-    archival : boolean
-        Reporting the source as an archival source (i.e. no upperlimit).
-    archival_comment : str
-        Comment on archival source. Required if archival is True.
-    instrument_ids : List[int]
-        Instrument IDs to restrict photometry to.
-    stream_ids : List[int]
-        Stream IDs to restrict photometry to.
-    parent_session : `sqlalchemy.orm.session.Session`
-        Database session.
-    """
-    print(instrument_ids)
-    print(stream_ids)
-
-    if parent_session is None:
-        if Session.registry.has():
-            session = Session()
-        else:
-            session = Session(bind=DBSession.session_factory.kw["bind"])
-    else:
-        session = parent_session
-
-    user = session.scalar(sa.select(User).where(User.id == user_id))
-
-    flow = Flow()
 
-    try:
-        # for now we limit it to instruments and filters we have mapped to TNS
-        instruments = session.scalars(
-            Instrument.select(user).where(
-                sa.func.lower(Instrument.name).in_(list(TNS_INSTRUMENT_IDS.keys()))
-            )
-        ).all()
-        if len(instruments) == 0:
-            raise ValueError('No instrument with known IDs available.')
-
-        tnsrobot = session.scalars(
-            TNSRobot.select(user).where(TNSRobot.id == tnsrobot_id)
-        ).first()
-        if tnsrobot is None:
-            raise ValueError(f'No TNSRobot available with ID {tnsrobot_id}')
-
-        altdata = tnsrobot.altdata
-        if not altdata:
-            raise ValueError('Missing TNS information.')
-        if 'api_key' not in altdata:
-            raise ValueError('Missing TNS API key.')
-
-        if archival is True:
-            if len(archival_comment) == 0:
-                raise ValueError(
-                    'If source flagged as archival, archival_comment is required'
+@check_loaded(logger=log)
+def service(*args, **kwargs):
+    while True:
```

Comment:
This is quite a long piece of code. Would it make sense to break it into functions? Something along the lines of:

```
while True:
    next_task = fetch_task()
    if next_task is None:
        time.sleep(5)
    else:
        service_task(next_task)
```

Reply:
Yes! Exactly what I've been doing this morning. I'm trying to avoid redundant pieces of code too.

Reply:
Done!